networkx/algorithms/community/centrality.py

```diff
@@ -120,32 +136,63 @@ def girvan_newman(G, most_valuable_edge=None):
     result can be depicted as a dendrogram.
 
     """
+
     # If the graph is already empty, simply return its connected
     # components.
     if G.number_of_edges() == 0:
         yield tuple(nx.connected_components(G))
         return
+
     # If no function is provided for computing the most valuable edge,
     # use the edge betweenness centrality.
-    if most_valuable_edge is None:
-
-        def most_valuable_edge(G):
-            """Returns the edge with the highest betweenness centrality
-            in the graph `G`.
-
-            """
-            # We have guaranteed that the graph is non-empty, so this
-            # dictionary will never be empty.
-            betweenness = nx.edge_betweenness_centrality(G)
-            return max(betweenness, key=betweenness.get)
+    component_wise_computing = False
+    fast_betweenness_centrality = False
+    if most_valuable_edge is None and most_valuable_edge_metric is None:
+        fast_betweenness_centrality = True
+    elif most_valuable_edge_metric:
+        component_wise_computing = True
 
     # The copy of G here must include the edge weight data.
     g = G.copy().to_undirected()
+
     # Self-loops must be removed because their removal has no effect on
     # the connected components of the graph.
     g.remove_edges_from(nx.selfloop_edges(g))
+
+    if fast_betweenness_centrality:
+        edge_betweenness = {}  #
+        edge_sources = {}  # edge_sources[edge] = [nodes which has shortest paths that go through edge]
+        node_contributions = {}  # contribution of a node to each edge's centrality
+
+        # Initial full betweenness calculation
+        for node in G.nodes():
+            S, P, sigma, _ = _single_source_shortest_path_basic(G, node)
+            node_contributions[node] = _compute_node_contributions(S, P, sigma)
+            for edge, score in node_contributions[node].items():
+                if edge not in edge_betweenness:
+                    edge_betweenness[edge] = 0.0
+                edge_betweenness[edge] += score
+
+                if edge not in edge_sources:
+                    edge_sources[edge] = []
```

Comment:
Removing elements from a list is expensive. We can make `edge_sources[edge]` a set instead to speed it up.